bundles/org.openhab.binding.froniuswattpilot/src/main/java/org/openhab/binding/froniuswattpilot/internal/FroniusWattpilotHandler.java

```diff
@@ -0,0 +1,311 @@
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.froniuswattpilot.internal;
+
+import static org.openhab.binding.froniuswattpilot.internal.FroniusWattpilotBindingConstants.*;
+
+import java.io.IOException;
+import java.net.NoRouteToHostException;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.io.EofException;
+import org.openhab.core.library.types.DecimalType;
+import org.openhab.core.library.types.OnOffType;
+import org.openhab.core.library.types.QuantityType;
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.library.unit.Units;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.ThingUID;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.florianhotze.wattpilot.WattpilotClient;
+import com.florianhotze.wattpilot.WattpilotClientListener;
+import com.florianhotze.wattpilot.WattpilotInfo;
+import com.florianhotze.wattpilot.WattpilotStatus;
+import com.florianhotze.wattpilot.commands.SetChargingCurrentCommand;
+import com.florianhotze.wattpilot.commands.SetChargingModeCommand;
+import com.florianhotze.wattpilot.commands.SetEnforcedChargingStateCommand;
+import com.florianhotze.wattpilot.commands.SetSurplusPowerThresholdCommand;
+import com.florianhotze.wattpilot.dto.ChargingMode;
+import com.florianhotze.wattpilot.dto.EnforcedChargingState;
+
+/**
+ * The {@link FroniusWattpilotHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Florian Hotze - Initial contribution
+ */
+@NonNullByDefault
+public class FroniusWattpilotHandler extends BaseThingHandler implements WattpilotClientListener {
+    private final Logger logger = LoggerFactory.getLogger(FroniusWattpilotHandler.class);
+    private final WattpilotClient client;
+
+    private @Nullable CompletableFuture<?> awaitDisconnect;
+
+    private @Nullable FroniusWattpilotConfiguration config;
+
+    public FroniusWattpilotHandler(Thing thing, HttpClient httpClient) {
+        super(thing);
+        client = new WattpilotClient(httpClient);
+        client.addListener(this);
```

Comment:
I'm wondering if it's a problem that there's no `removeListener` available? Usually the handler would add itself as a listener in `initialize` and remove itself in `dispose`. I guess by subscribing outside of these lifecycle methods, we could receive events even when not initialized, causing the handler to update channels and Thing status?

Reply:
Hmm, the question is what is the lifecycle of the handler here.
If the handler instance is removed after calling dispose, we don't need to remove the listteners as the client instance will also be removed together with the handler.
My wattpilot4j library and the binding itself follow the structure of the Harmony Hub binding, I am pretty sure there no listener is removed.

Reply:
Internally the listeners are stored in a HashSet, so they cannot be added multiple times.
Once you call disconnect, onDisconnect will trigger once and then you are disconnected and won’t receive any updates.